File: lib/domain/fen/fen_codec.dart
import '../models/board.dart';
import '../models/board_point.dart';
import '../models/piece.dart';
import '../models/piece_type.dart';
import '../models/side.dart';
import '../rules/board_geometry.dart';

class FenParseException implements Exception {
  final String message;
  FenParseException(this.message);

  @override
  String toString() => 'FenParseException: $message';
}

class DecodedFen {
  final Board board;
  final Side sideToMove;
  final int halfmoveClock;
  final int fullmoveNumber;

  const DecodedFen({
    required this.board,
    required this.sideToMove,
    this.halfmoveClock = 0,
    this.fullmoveNumber = 1,
  });
}

/// Xiangqi FEN, using the same piece-letter convention as Pikafish/UCCI
/// engines (K/A/B/N/R/C/P for General/Advisor/Elephant/Horse/Chariot/Cannon/
/// Soldier), so this codec's output can be fed to a Pikafish-backed backend
/// unchanged in a later phase. Uppercase = Red, lowercase = Black.
///
/// Format: `<board> <sideToMove> - - <halfmoveClock> <fullmoveNumber>`
/// The board field lists ranks top-to-bottom (Black's back rank, i.e. row 9,
/// first; Red's back rank, row 0, last), matching standard FEN row order.
/// `w` means Red to move, `b` means Black to move (Red is the first-moving
/// side, playing the role chess assigns to White).
class FenCodec {
  FenCodec._();

  static const Map<PieceType, String> _redLetters = {
    PieceType.general: 'K',
    PieceType.advisor: 'A',
    PieceType.elephant: 'B',
    PieceType.horse: 'N',
    PieceType.chariot: 'R',
    PieceType.cannon: 'C',
    PieceType.soldier: 'P',
  };

  static final Map<String, PieceType> _letterToType = {
    for (final entry in _redLetters.entries) entry.value: entry.key,
  };

  static String encode(
    Board board,
    Side sideToMove, {
    int halfmoveClock = 0,
    int fullmoveNumber = 1,
  }) {
    final rankStrings = <String>[];
    for (var row = BoardGeometry.rows - 1; row >= 0; row--) {
      final buffer = StringBuffer();
      var emptyRun = 0;
      for (var col = 0; col < BoardGeometry.columns; col++) {
        final piece = board.pieceAt(BoardPoint(row, col));
        if (piece == null) {
          emptyRun++;
          continue;
        }
        if (emptyRun > 0) {
          buffer.write(emptyRun);
          emptyRun = 0;
        }
        final letter = _redLetters[piece.type]!;
        buffer.write(piece.side == Side.red ? letter : letter.toLowerCase());
      }
      if (emptyRun > 0) buffer.write(emptyRun);
      rankStrings.add(buffer.toString());
    }
    final boardField = rankStrings.join('/');
    final sideField = sideToMove == Side.red ? 'w' : 'b';
    return '$boardField $sideField - - $halfmoveClock $fullmoveNumber';
  }

  static DecodedFen decode(String fen) {
    final parts = fen.trim().split(RegExp(r'\s+'));
    if (parts.length < 2) {
      throw FenParseException('FEN must have at least board and side-to-move fields');
    }
    final boardField = parts[0];
    final sideField = parts[1];

    final rankStrings = boardField.split('/');
    if (rankStrings.length != BoardGeometry.rows) {
      throw FenParseException(
          'Expected ${BoardGeometry.rows} ranks separated by "/", got ${rankStrings.length}');
    }

    final grid = List.generate(
      BoardGeometry.rows,
      (_) => List<Piece?>.filled(BoardGeometry.columns, null),
    );

    for (var i = 0; i < rankStrings.length; i++) {
      final row = BoardGeometry.rows - 1 - i; // first FEN rank = top = row 9
      var col = 0;
      for (final ch in rankStrings[i].split('')) {
        final digit = int.tryParse(ch);
        if (digit != null) {
          col += digit;
          continue;
        }
        final type = _letterToType[ch.toUpperCase()];
        if (type == null) throw FenParseException('Unknown piece letter "$ch"');
        if (col >= BoardGeometry.columns) {
          throw FenParseException('Rank "${rankStrings[i]}" overflows board width');
        }
        final side = ch == ch.toUpperCase() ? Side.red : Side.black;
        grid[row][col] = Piece(type, side);
        col++;
      }
      if (col != BoardGeometry.columns) {
        throw FenParseException(
            'Rank "${rankStrings[i]}" has $col squares, expected ${BoardGeometry.columns}');
      }
    }

    Side side;
    if (sideField == 'w') {
      side = Side.red;
    } else if (sideField == 'b') {
      side = Side.black;
    } else {
      throw FenParseException('Side-to-move field must be "w" or "b", got "$sideField"');
    }

    final halfmove = parts.length > 4 ? int.tryParse(parts[4]) ?? 0 : 0;
    final fullmove = parts.length > 5 ? int.tryParse(parts[5]) ?? 1 : 1;

    return DecodedFen(
      board: Board.fromGrid(grid),
      sideToMove: side,
      halfmoveClock: halfmove,
      fullmoveNumber: fullmove,
    );
  }
}

File: backend/src/xiangqi/fenCodec.ts
import { Board } from './board';
import { BOARD_COLUMNS, BOARD_ROWS } from './boardGeometry';
import { Piece, piece as makePiece } from './piece';
import { PieceType } from './pieceType';
import { Side } from './side';

/**
 * Same convention as the client's `lib/domain/fen/fen_codec.dart` and this
 * backend's own Pikafish wrapper (`utils/fen.ts` validates the same shape):
 * K/A/B/N/R/C/P uppercase = Red, lowercase = Black; board field lists ranks
 * top-to-bottom (row 9 first); 'w' = Red to move, 'b' = Black.
 */

const RED_LETTERS: Record<PieceType, string> = {
  general: 'K',
  advisor: 'A',
  elephant: 'B',
  horse: 'N',
  chariot: 'R',
  cannon: 'C',
  soldier: 'P',
};

const LETTER_TO_TYPE: Record<string, PieceType> = Object.fromEntries(
  Object.entries(RED_LETTERS).map(([type, letter]) => [letter, type as PieceType]),
);

export interface DecodedFen {
  board: Board;
  sideToMove: Side;
  halfmoveClock: number;
  fullmoveNumber: number;
}

export function encodeFen(board: Board, sideToMove: Side, halfmoveClock = 0, fullmoveNumber = 1): string {
  const rankStrings: string[] = [];
  for (let row = BOARD_ROWS - 1; row >= 0; row--) {
    let buffer = '';
    let emptyRun = 0;
    for (let col = 0; col < BOARD_COLUMNS; col++) {
      const pc = board.pieceAt({ row, col });
      if (pc === null) {
        emptyRun++;
        continue;
      }
      if (emptyRun > 0) {
        buffer += emptyRun;
        emptyRun = 0;
      }
      const letter = RED_LETTERS[pc.type];
      buffer += pc.side === 'red' ? letter : letter.toLowerCase();
    }
    if (emptyRun > 0) buffer += emptyRun;
    rankStrings.push(buffer);
  }
  const boardField = rankStrings.join('/');
  const sideField = sideToMove === 'red' ? 'w' : 'b';
  return `${boardField} ${sideField} - - ${halfmoveClock} ${fullmoveNumber}`;
}

export function decodeFen(fen: string): DecodedFen {
  const parts = fen.trim().split(/\s+/);
  if (parts.length < 2) throw new Error('FEN must have at least board and side-to-move fields');
  const [boardField, sideField] = parts;

  const rankStrings = boardField!.split('/');
  if (rankStrings.length !== BOARD_ROWS) {
    throw new Error(`Expected ${BOARD_ROWS} ranks separated by "/", got ${rankStrings.length}`);
  }

  const grid: (Piece | null)[][] = Array.from({ length: BOARD_ROWS }, () =>
    Array<Piece | null>(BOARD_COLUMNS).fill(null),
  );

  for (let i = 0; i < rankStrings.length; i++) {
    const row = BOARD_ROWS - 1 - i; // first FEN rank = top = row 9
    let col = 0;
    for (const ch of rankStrings[i]!) {
      const digit = Number.parseInt(ch, 10);
      if (!Number.isNaN(digit)) {
        col += digit;
        continue;
      }
      const type = LETTER_TO_TYPE[ch.toUpperCase()];
      if (type === undefined) throw new Error(`Unknown piece letter "${ch}"`);
      if (col >= BOARD_COLUMNS) throw new Error(`Rank "${rankStrings[i]}" overflows board width`);
      const side: Side = ch === ch.toUpperCase() ? 'red' : 'black';
      grid[row]![col] = makePiece(type, side);
      col++;
    }
    if (col !== BOARD_COLUMNS) {
      throw new Error(`Rank "${rankStrings[i]}" has ${col} squares, expected ${BOARD_COLUMNS}`);
    }
  }

  let side: Side;
  if (sideField === 'w') side = 'red';
  else if (sideField === 'b') side = 'black';
  else throw new Error(`Side-to-move field must be "w" or "b", got "${sideField}"`);

  const halfmoveClock = parts.length > 4 ? Number.parseInt(parts[4]!, 10) || 0 : 0;
  const fullmoveNumber = parts.length > 5 ? Number.parseInt(parts[5]!, 10) || 1 : 1;

  return { board: Board.fromGrid(grid), sideToMove: side, halfmoveClock, fullmoveNumber };
}
